src/dag_data.hpp

```diff
@@ -0,0 +1,254 @@
+           "Data Vector must be of equal size to current count.");
+    for (size_t i = 0; i < GetCount(); i++) {
+      data_vec_[i] = data_vec[i];
+    }
+  }
+
+  // Value to assign to newly added elements.
+  std::optional<DataType> HasDefaultValue() { return default_value_.has_value(); }
+  DataType GetDefaultValue() {
+    Assert(default_value_.has_value(), "Cannot be called when DefaultValue not set.");
+    return default_value_.value();
+  }
+  void SetDefaultValue(const DataType default_value) { default_value_ = default_value; }
+
+  // Fill all cells with given value.
+  void Fill(const DataType fill_value) {
+    for (size_t i = 0; i < GetPaddedCount(); i++) {
+      data_vec_[i] = fill_value;
+    }
+  }
+  void FillWithDefault() {
+    Assert(default_value_.has_value(), "Cannot be called when DefaultValue not set.");
+    Fill(default_value_.value());
+  }
+
+  // ** Maintanence
+
+  virtual void Resize(std::optional<const Reindexer> reindexer = std::nullopt,
+                      std::optional<const size_t> explicit_alloc = std::nullopt) {
+    Failwith("Unspecified DAGData cannot call base Resize without explicit count.");
+  }
+  virtual void Resize(const size_t new_count,
+                      std::optional<const Reindexer> reindexer = std::nullopt,
+                      std::optional<const size_t> explicit_alloc = std::nullopt) {
+    size_t old_count = GetCount();
+    SetCount(new_count);
+    if (GetPaddedCount() > GetAllocCount()) {
+      SetAllocCount(size_t(ceil(double(GetPaddedCount()) * resizing_factor_)));
+      if (explicit_alloc.has_value()) {
+        Assert(explicit_alloc.value() >= GetCount(),
+               "Attempted to reallocate space smaller than count.");
+        SetAllocCount(explicit_alloc.value() + GetSpareCount());
+      }
+      data_vec_.conservativeResize(GetAllocCount());
+    }
+    data_vec_.conservativeResize(GetPaddedCount());
+    if (default_value_.has_value()) {
+      for (size_t i = old_count; i < GetPaddedCount(); i++) {
+        data_vec_[i] = default_value_.value();
+      }
+    }
+    if (reindexer.has_value()) {
+      Reindex(reindexer.value(), new_count);
+    }
+  }
+  void Resize(Resizer &resizer,
+              std::optional<const Reindexer> reindexer = std::nullopt) {
+    resizer.ApplyResizeToEigenVector<VectorType, DataType>(data_vec_, default_value_);
+    SetCount(resizer.new_count_);
+    SetAllocCount(resizer.new_alloc_);
+    if (reindexer.has_value()) {
+      Reindex(reindexer.value(), resizer.new_count_);
+    }
+  }
+
+  virtual void Reindex(const Reindexer &reindexer, const size_t length) {
+    Reindexer::ReindexInPlace<VectorType, DataType>(data_vec_, reindexer, length);
+  }
+
+ protected:
+  // Data vector
+  VectorType data_vec_;
+  size_t data_per_element_ = DataPerElement;
+  // Non-ownership DAG reference.
+  // Optional: Used for resizing data vectors, otherwise explicit arguments can
+  // be used.
+  GPDAG *dag_ = nullptr;
```

Comment:
My though process here is that it contains the reference to the `DAG`, so you can simply call the `DAGData` to resize, and it will lookup the DAG node or edges, and resize itself accordingly.  But if you eliminated this you could simply pass the DAG as an argument when resizing.  Or just the pass edge or node length explicitly, but at that point its functionality more or less just a padded data vector.

Reply:
I like your alternate suggestions. And there's no problem with having a class for a padded data vector! Simple and composable are good.